### DISCUSSION TITLE
Cursor Pagination

### DISCUSSION THREAD

**Discussion by StephenStrickland** [alirezanet/Gridify#102] [Q&A]
Love the project.

I'm looking to implement cursor based pagination, any easy way to do that currently with Gridify?
[edit: spelling]

**Comment by alirezanet** [comment]
Interesting question, maybe soon I would provide some built-in feature for it but right now gridify doesn't support it directly

**Reply by StephenStrickland** [reply]
Would you be interested a PR or a RFC to talk through potential implementation?

**Reply by alirezanet** [reply]
ofc would be nice and helpful !

**Reply by cmenzi** [reply]
@StephenStrickland Did you already started with it? I've tried it, but it's not that easy., 

One issue is that a lot of extension methods relies on `IGridifyQuery`, which already derives from `IGridifyPagination`. So, we would need to re-design this heavily, I guess.

I've introduced a `IGridifyCursorPagination` and also some extension methods on it 

```
IQueryable<T> ApplyCursorPaging<T>(this IQueryable<T> query, IGridifyCursorPagination? gridifyCursorPagination)
```

```csharp
public interface IGridifyCursorPagination : IGridifyOrdering
{
    int Limit { get; set; }

    string Cursor { get; set; }
}
```

But then methods like `Gridify`, `GridifyQueryable`, `GridifyAsync`, `GridifyQueryableAsync` not working anymore, as they need a `IGridifyQuery`, which as I mentioned already implements `IGridifyPagination`.

I think we need to include `IGridifyOrdering` as we need remember the ordering of the columns. Not sure we also need to remember the filter. We also need include ordering of the id always `id asc` as last order statement. 

Open Questions:
- How to handle Guids as Ids?
- What if the user specify ordering `id desc` by himself?
- How to handle different namings of the id columns. We need to order by some determinist id or compound key?

The idea would then be to base64 encode the cursor with the orderings and the `nextToken` and return as `Cursor` property.

I've found this good article here: https://medium.com/@george_16060/cursor-based-pagination-with-arbitrary-ordering-b4af6d5e22db

Other inputs:
- https://docs.gitlab.com/ee/api/rest/#keyset-based-pagination
- https://github.com/mrahhal/MR.AspNetCore.Pagination

Maybe we could go the way as GitLab by adding `pagination=keyset` to the model. But I don't really like that.

**Comment by cmenzi** [comment]
@StephenStrickland @alirezanet You can find my first implementation ideas here. It's not really ready yet, but maybe to start some dicusssions

https://github.com/alirezanet/Gridify/compare/master...cmenzi:Gridify:feature/cursor-based

**Reply by alirezanet** [reply]
Hi @cmenzi, after a quick review, I noticed substantial changes to contracts (interfaces), potentially causing numerous breaking changes. Such extensive alterations typically require a new major release. However, to maintain user-friendly migration, we could consider a way to support cursor pagination with fewer modifications.

I haven't delved into details yet, but introducing additional interfaces, like IGridifyCursorQuery, alongside the existing ones might be a better approach. This would allow for an alternative implementation while preserving the current interfaces. Overloads in the extension methods could further aid the transition.
what do you think?

**Reply by alirezanet** [reply]
Or if we really want to accept this amount of changes for version 3, I would suggest finding a better way (like a configurable setting) to switch between Cursor pagination and Offset pagination.

since we are mostly dealing with parsing strings ... it is possible.

**Reply by cmenzi** [reply]
@alirezanet Yes, I also don't like those breaking changes. Yeah, that could be an option.

Option 1) You mean keep `IGridifyQuery` as it is and try to add side-by-side implementation with `IGridifyCursorQuery` and necessary extentions methods.

Option 2) Yes, for a version 3 we could do that. Question is if how would the model `GridifyQuery` look like? Having all properties for both paging method in it? Or having different models? ...

I would try to do another suggestion - less intrusive :-) - for option 1.

**Reply by alirezanet** [reply]
Option 1- Yes this way we don't break anything.
Option 2- anyway we need to improve the library in v3 ... not sure yet ... needs some investigations

**Reply by cmenzi** [reply]
Here 2nd attempt with option 1: https://github.com/alirezanet/Gridify/compare/master...cmenzi:Gridify:feature/cursor-option-1?expand=1

Still need to do some refactorings. But if the direction now is ok, I will raise a the PR after I've done all the clean ups, docs, ...

**Reply by alirezanet** [reply]
Nice! 🚀 Looks much better.
would be nice if you can also add some documentation explaining how it works and what are the limitations in the docs

**Reply by alirezanet** [reply]
@StephenStrickland can you test this branch and let me know if it works for you?
